README.md
# Flask application

Generate a Python application using the [Flask](http://flask.pocoo.org/)
framework by clicking the next button:

[<img src="https://images.atomist.com/button/create-project.png" width="267" alt="Set Started with Atomist"/>](https://api.atomist.com/v1/projects/generators/f57c6427-f2de-4cf4-931a-b1221a9cd08a)

This will ask you to authorize the [Atomist](https://www.atomist.com/)
application in your GitHub organisation/account then will ask you to provide
a set of parameters to generate the Flask application skeleton.

Once the project is generated in your GitHub organisation/account, simply
checkout the newly generated project and follow its README to get started.

You can also enhance your new project, or an existing one, by going through the
editors below:

[![Add support for building Docker images from your Flask project](https://images.atomist.com/button/add-docker.png)](https://api.atomist.com/v1/projects/editors/47aff86f-afcb-4b6a-be33-4f4682cc80bc)

---
Created by Atomist. Need Help? [Join our Slack team](https://join.atomist.com/)

.atomist/CHANGELOG.md
# Change Log

Changes about the Flask generator and editors.

## [1.1.2] - Unreleased
### Added
- Changelog

## [1.1.1] - 2016-10-31
### Changed
- Improved parameter descriptions of the `SetProjectReadme` editor

## [1.1.0] - 2016-10-31
### Changed
- Made parameter' patterns more consistent
- Improved parameter descriptions and input validation messages

## [1.0.15] - 2016-10-31
### Changed
- Getting project name so we can inject it into Dockerfile and README

## [1.0.11] - 2016-10-31
### Changed
- Registered `AddDockerfile` editor as a button

## [1.0.10] - 2016-10-30
### Added
- Added `AddTravis` editor

## [1.0.9] - 2016-10-30
### Removed
- Old style pom resources in favor of `manifest.yml`

flask_service/tests/test_swagger.py
# -*- coding: utf-8 -*-
from flask_service.swagger import spec
from flask_service import __version__


def test_document_meta():
    api = spec.to_dict()
    assert api['info']['title'] == 'flask_service'
    assert api['info']['version'] == __version__
    assert api['info']['description'] == 'My application'


def test_document_health_endpoint():
    api = spec.to_dict()
    assert '/health' in api['paths']
    assert 'get' in api['paths']['/health']
    assert '200' in api['paths']['/health']['get']['responses']
    assert 'string' in api['paths']['/health']['get']['responses']['200']['schema']
    assert '503' in api['paths']['/health']['get']['responses']
    assert 'string' in api['paths']['/health']['get']['responses']['503']['schema']


def test_document_status_endpoint():
    api = spec.to_dict()
    assert '/status' in api['paths']
    assert 'get' in api['paths']['/status']
    assert '200' in api['paths']['/status']['get']['responses']
    assert 'string' in api['paths']['/status']['get']['responses']['200']['schema']


def test_document_my_app_endpoint():
    api = spec.to_dict()
    assert '/' in api['paths']
    assert 'get' in api['paths']['/']
    assert '200' in api['paths']['/']['get']['responses']
    assert 'string' in api['paths']['/']['get']['responses']['200']['schema']

flask_service/my_app/tests/conftest.py
# -*- coding: utf-8 -*-
import pytest

from flask_service import create_app


@pytest.fixture
def app():
    app = create_app()
    return app


flask_service/tests/test_main_views.py
# -*- coding: utf-8 -*-
from flask import url_for


def test_swagger_endpoint(client):
    res = client.get(url_for('main_app.api'))
    assert isinstance(res.data, dict)


flask_service/runserver.py
# -*- coding: utf-8 -*-
import cherrypy
import click

from flask_service import create_app


app = create_app()


def run_dev():
    """
    Run the application using the built-in server.
    """
    app.config.from_object("flask_service.settings_dev")
    app.run()


def run_prod():
    """
    Run the application in the well-established CherryPy
    WSGI server. That server is multithreaded and
    has been production ready for many years.

    .. seealso: http://cherrypy.org/
    """
    app.config.from_object("flask_service.settings_prod")
    cherrypy.config.update({
        "environment": "production",
        "server.socket_host": "0.0.0.0",
        "server.socket_port": 5000,
        "log.screen": True
    })
    cherrypy.tree.graft(app)
    cherrypy.engine.signals.subscribe()
    cherrypy.engine.start()
    cherrypy.engine.block()


@click.command()
@click.option('--dev', 'mode', flag_value='dev')
@click.option('--prod', 'mode', flag_value='prod')
def run(mode):
    """
    Entrypoint to run the application from the
    command line, choosing between the `dev` and
    `prod` modes.
    """
    if mode == 'dev':
        run_dev()
    elif mode == 'prod':
        run_prod()


if __name__ == '__main__':
    run()


flask_service/swagger.py
# -*- coding: utf-8 -*-
from apispec import APISpec

__all__ = ['spec']


spec = APISpec(
    title='flask_service',
    version='0.0.1',
    info=dict(
        description='My application'
    ),
    plugins=('apispec.ext.marshmallow',)
)


spec.add_path(
    path='/',
    operations=dict(
        get=dict(
            responses={
                '200': {
                    'schema': 'string'
                }
            }
        )
    )
)


spec.add_path(
    path='/health',
    description='Indicates the health of the application',
    operations=dict(
        get=dict(
            responses={
                '200': {
                    'schema': 'string'
                },
                '503': {
                    'schema': 'string',
                    'description': 'The application is not available'
                }
            }
        )
    )
)


spec.add_path(
    path='/status',
    description='Indicates the live status of the application',
    operations=dict(
        get=dict(
            responses={
                '200': {
                    'schema': 'string'
                }
            }
        )
    )
)